src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -1565,6 +1566,18 @@ internal new NamedTypeSymbol GetSpecialType(SpecialType specialType)
             return result;
         }
 
+        /// <summary>
+        /// Cache of T to Nullable&lt;T&gt;.
+        /// </summary>
+        private ImmutableSegmentedDictionary<TypeSymbol, NamedTypeSymbol> _typeToNullableVersion = ImmutableSegmentedDictionary<TypeSymbol, NamedTypeSymbol>.Empty;
+
+        internal NamedTypeSymbol GetOrCreateNullableType(TypeSymbol typeArgument)
```

Comment:
@CyrusNajmabadi This feels to general, We probably should limit the set of possible ```typeArgument``` values here. Perhaps with an assert.

Reply:
What would you like this limited to?

Reply:
> What would you like this limited to?

It looks like the intent to cache values for a very specific set of underlying types. Let's limit to that set

Reply:
Done in followup PR.